File: Q36-SortStack.cpp
//
//  Q36-SortStack.cpp
//  Algorithm-Linux
//
//  Created by Sanqiang Zhao on 12/10/12.
//  Copyright (c) 2012 Sanqiang Zhao. All rights reserved.
//
/*
 Stack<int> *st = new Stack<int>();
 st->push(3);
 st->push(2);
 st->push(1);
 st->push(4);
 st->push(5);
 st->print();
 
 Stack<int> *result = sortStack2(st);
 result->print();
 
 //sortStack(st);
 //st->print();
*/
#include "Q36-SortStack.h"

void sortStack(Stack<int> *&s)
{
    Stack<int> *helper = new Stack<int>();
    while (1) {
        int val = s->pop();
        while (s->size > 0 && s->peek() <= val) {
            helper->push(s->pop());
        }
        s->push(val);
        if (helper->size == 0) {
            return;
        }
        while (helper->size > 0) {
            s->push(helper->pop());
        }
    }
}

Stack<int> * sortStack2(Stack<int> *s)
{
    Stack<int> *helper = new Stack<int>();
    while (s->size > 0) {
        int val = s->pop();
        while (helper->size > 0 && helper->peek() <= val) {
            s->push(helper->pop());
        }
        helper->push(val);
    }
    
    return helper;
}

File: Algorithm-Java/src/CareerCup/StacksAndQueues/Q3_6_SortStack.java
/**
 * Sanqiang Zhao Www.131X.Com Jan 4, 2013
 */
package CareerCup.StacksAndQueues;

import java.util.Stack;

public class Q3_6_SortStack {

    private Stack<Integer> Helper = new Stack<>();

    public void sort(Stack<Integer> stack) {
        while (true) {
            while (Helper.isEmpty() || (!stack.isEmpty() && Helper.peek() <= stack.peek())) {
                Helper.push(stack.pop());
            }
            if (stack.isEmpty()) {
                break;
            } else {
                int val = stack.pop();
                while (!Helper.isEmpty()) {
                    stack.push(Helper.pop());
                }
                stack.push(val);
            }
        }
        while (!Helper.isEmpty()) {
            stack.push(Helper.pop());
        }
    }

    public static void main(String[] args) {
        Stack<Integer> stack = new Stack<>();
        stack.push(2);
        stack.push(5);
        stack.push(4);
        stack.push(3);
        stack.push(6);
        stack.push(1);
        Q3_6_SortStack q36 = new Q3_6_SortStack();
        q36.sort(stack);
        for (int i = 0; i < stack.size(); i++) {
            System.out.println(stack.get(i));
        }
    }
}
